File: go/handlers/todo.go
package handlers

import (
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/nunows/todo-api/go/models"
)

//Common web error
var (
	invalidParamMsg = models.Status{Success: false, Msg: "Error: Invalid Parameter."}
	missingParamMsg = models.Status{Success: false, Msg: "Error: Missing required Parameter."}
	notFoundMsg     = models.Status{Success: false, Msg: "Error: Not found."}
)

type TodoHandler struct {
	Db models.TodoDb
}

func (th *TodoHandler) GetAll(c *gin.Context) {
	todos, err := th.Db.GetAll()
	if err != nil {
		c.JSON(500, models.Status{Success: false, Msg: err.Error()})
		return
	}
	c.JSON(200, todos)
}

func (th *TodoHandler) Get(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(400, invalidParamMsg)
		return
	}

	todo, err := th.Db.Get(id)
	if err != nil {
		c.JSON(404, notFoundMsg)
	} else {
		c.JSON(200, todo)
	}

}

func (th *TodoHandler) Create(c *gin.Context) {
	var todo models.Todo
	err := c.BindJSON(&todo)
	if err != nil {
		c.JSON(500, missingParamMsg)
		return
	}
	err = th.Db.Insert(&todo)
	if err != nil {
		c.JSON(500, models.Status{Success: false, Msg: err.Error()})
	} else {
		c.JSON(200, models.Status{Success: true, Msg: "Todo created."})
	}
}

func (th *TodoHandler) Update(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(400, invalidParamMsg)
		return
	}
	var todo models.Todo

	err = c.BindJSON(&todo)
	if err != nil {
		c.JSON(500, missingParamMsg)
		return
	}

	err = th.Db.Update(id, &todo)
	if err != nil {
		c.JSON(500, models.Status{Success: false, Msg: err.Error()})
	} else {
		c.JSON(200, models.Status{Success: true, Msg: "Todo updated."})
	}
}

func (th *TodoHandler) Delete(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(400, invalidParamMsg)
		return
	}

	err = th.Db.Delete(id)
	if err != nil {
		c.JSON(404, models.Status{Success: false, Msg: err.Error()})
	} else {
		c.JSON(200, models.Status{Success: true, Msg: "Todo updated."})
	}

}

File: nodejs/handlers/todo.js
var Joi = require('joi');

var tododb = require('../db/tododb');

module.exports = {
    getAll: {
        handler: function(request, reply) {
            tododb.getAll(function(data) {
                if(data){
                    reply(data);
                } else {
                    reply({"success": false, "msg": "Error: Empty."});
                }
            });
        }
    },
    get: {
        handler: function(request, reply) {
            id = request.params.id;
            tododb.get(id, function(data) {
                if(data){
                    reply(data);
                } else {
                    reply({"success": false, "msg": "Error: Todo not found."});
                }
            });
        }, validate: {
            params: { id: Joi.number().integer().min(1) }
        }
    },
    create: {
        handler: function(request, reply) {
            var todo = {
                name: request.payload.name,
                done: request.payload.done
            };

            tododb.insert(todo, function(data) {
                reply({"success": true, "msg": "Todo created."});
            });

        }, validate: {
            payload: {
                name: Joi.string().required(),
                done: Joi.boolean().required()
            }
        }
    },
    update: {
        handler: function(request, reply) {
            id = request.params.id;
            var todo = {
                name: request.payload.name,
                done: request.payload.done
            };
            tododb.update(id, todo, function(data) {
                if(data){
                    reply({"success": true, "msg": "Todo updated."});
                } else {
                    reply({"success": false, "msg": "Error: Updating Todo."});
                }
            });
        }, validate: {
            params: { id: Joi.number().integer().min(1) },
            payload: {
                name: Joi.string().required(),
                done: Joi.boolean().required()
            }
        }
    },
    delete: {
        handler:function(request, reply) {
            id = request.params.id;
            tododb.delete(id, function(data) {
                if(data){
                    reply({"success": true, "msg": "Todo deleted."});
                } else {
                    reply({"success": false, "msg": "Error: Deleting Todo."});
                }
            });
        }, validate: {
            params: { id: Joi.number().integer().min(1) }
        }
    }
};
